src/main/java/cloud/eppo/BaseEppoClient.java

```diff
@@ -156,33 +165,55 @@ protected EppoValue getTypedAssignment(
     }
 
     if (assignedValue != null && assignmentLogger != null && evaluationResult.doLog()) {
-      String allocationKey = evaluationResult.getAllocationKey();
-      String experimentKey =
-          flagKey
-              + '-'
-              + allocationKey; // Our experiment key is derived by hyphenating the flag key and
-      // allocation key
-      String variationKey = evaluationResult.getVariation().getKey();
-      Map<String, String> extraLogging = evaluationResult.getExtraLogging();
-      Map<String, String> metaData = buildLogMetaData(config.isConfigObfuscated());
-
-      Assignment assignment =
-          new Assignment(
-              experimentKey,
-              flagKey,
-              allocationKey,
-              variationKey,
-              subjectKey,
-              subjectAttributes,
-              extraLogging,
-              metaData);
+
       try {
-        assignmentLogger.logAssignment(assignment);
+        String allocationKey = evaluationResult.getAllocationKey();
+        String experimentKey =
+            flagKey
+                + '-'
+                + allocationKey; // Our experiment key is derived by hyphenating the flag key and
+        // allocation key
+        String variationKey = evaluationResult.getVariation().getKey();
+        Map<String, String> extraLogging = evaluationResult.getExtraLogging();
+        Map<String, String> metaData = buildLogMetaData(config.isConfigObfuscated());
+
+        Assignment assignment =
+            new Assignment(
+                experimentKey,
+                flagKey,
+                allocationKey,
+                variationKey,
+                subjectKey,
+                subjectAttributes,
+                extraLogging,
+                metaData);
+
+        // Deduplication of assignment logging is possible by providing an `IAssignmentCache`.
+        // Default to true, only avoid logging if there's a cache hit.
+        boolean logAssignment = true;
+        AssignmentCacheEntry cacheEntry = AssignmentCacheEntry.fromVariationAssignment(assignment);
+        if (assignmentCache != null) {
+          try {
+            if (assignmentCache.hasEntry(cacheEntry).get()) {
```

Comment:
Given we call `get()` everywhere I'm thinking the cache interface should just be booleans and the use of completable futures can be done under the hood in concrete implementations

Reply:
You're right - this is actually a nice benefit of Java in that we can have functions that do the js equivalent of `await` without needing to also mark it as async.